---
title: "Analysis"
author: "Gary Wang"
date: "`r Sys.Date()`"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE, eval = FALSE)
```

## Introduction

This is the code for the analysis of the RNA-seq data. The reference files are downloaded from GENCODE "<https://www.gencodegenes.org/mouse/>". - The GTF file is the basic gene annotation (CHR) - The Fasta file is the primary assembly (GRCm39).

The code is largely based on the tutorial [here](https://star-protocols.cell.com/protocols/931#step-by-step-method-details) by Sanchis et al. and DESeq2 vignette by Love et al. [here](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). Refer to clusterProfiler vignette [here](http://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html) for pathway analysis.

## Optional: QC the input files

Before mapping the data, we can use fastqc and multiqc to generate QC metrics for the sequence files:

`fastqc --noextract --nogroup -o fastqc *.fastq.gz`

Then, within the fastqc folder (Docker needed):

```         
 multiqc .
```
## Data preprocessing
### Load the required libraries

```{r, echo = FALSE}
library(clusterProfiler)
library(AnnotationDbi)
library(org.Mm.eg.db)
library(DESeq2)
library(dendextend)
library(airway)
library(tidyr)
library(readxl)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(biomaRt)
library(edgeR)
library(Rsubread)
library(here)
library(RColorBrewer)
library(pheatmap)
library("factoextra")
library(ggrepel)
library(enrichplot)
#"E:/gary/Trm_2024/Analysis_Shuyi_Schneck"
```

### Define paths

```{r}
fastq_dir <- here("Seq_2024_04")
ref_dir <- here("genome", "GRCm39_M34")
meta_dir <- here("sample_info.csv")
analysis_out <- here("analysis")
```

### Preprocessing of reference data

We need to preprocess the reference genome and our fastq data with Rsubread to generate a count matrix. Only need to run once.

```{r, eval = FALSE}
buildindex(basename = "GRCm39_M34", reference = here(ref_dir, "GRCm39.primary_assembly.genome.fa.gz")) # Take around 30-50 min, depending on the computer.
```

### Read and align the fastq files to the genome

```{r, eval = FALSE}
reads1 <- list.files(path = here("Seq_2024_04"),
                     pattern = '*_1.fastq.gz$', full.names = TRUE)
reads2 <- list.files(path = here("Seq_2024_04"),
                     pattern = '*_2.fastq.gz$', full.names = TRUE)

align(index = "GRCm39_M34",
      readfile1 = reads1, readfile2 = reads2,
      input_format = "gzFASTQ",
      output_format = "BAM",
      nthreads = 12)
```

### Generate the count matrix

```{r, eval = FALSE}
bam_files <- list.files(here("Seq_2024_04"), 
                        pattern = "_1.fastq.gz.subread.BAM$", 
                        full.names = TRUE) #Remember to change pattern depending on input files
fc <- featureCounts(files = bam_files, 
                    annot.ext = here("genome", "GRCm39_M34","gencode.vM34.basic.annotation.gtf.gz"),
                    isGTFAnnotationFile = TRUE,
                    isPairedEnd = TRUE)
saveRDS(fc, file = here(analysis_out, "featureCounts.rds"))

raw_counts <- fc$counts
colnames(raw_counts) <- stringr::str_remove(colnames(raw_counts), pattern = "_1.fastq.gz.subread.BAM")
```

## DEG analysis
### Creating DESeq object
```{r, eval = FALSE}
# If reading count matrix from file
fc <- readRDS(here(analysis_out, "featureCounts.rds"))
raw_counts <- fc$counts
colnames(raw_counts) <- stringr::str_remove(colnames(raw_counts), pattern = "_1.fastq.gz.subread.BAM")
```

```{r, eval = FALSE}
## Read meta data
meta_data <- read.csv(meta_dir, stringsAsFactors = FALSE, header = TRUE)
meta_data$Type <- as.factor(meta_data$Type)
rownames(meta_data) <- meta_data$Sample # Help match the roles of raw count matrix

meta_data$Batch <- factor(meta_data$Batch)
colnames(raw_counts) <- c("Teff1","Teff2","Teff3","Trm-like1","Trm-like2","Trm-like3")
# For the sake of analysis, we simplify the gene names to exclude versions.
rownames(raw_counts) <- sapply(rownames(raw_counts), function(x) strsplit(x, split = "\\.")[[1]][1])
## Create DESeq object by merging count and metadata  
dds <- DESeqDataSetFromMatrix(countData = raw_counts,
                              colData = meta_data,
                              design = ~ Batch+Type)
```

```{r}
head(dds)
```
```{r, eval = FALSE}
# Filter the genes with low counts
keep <- rowSums(counts(dds) >= 3)>= min(table(meta_data$Type)) # Keep genes with at least 3 counts in at least 3 samples (smallest group size). 

dds <- dds[keep,]
```

### Perform statistical tests to identify DEG
```{r}
dds <- DESeq(dds)
# Make sure we select the reference level
dds$Type <- relevel(dds$Type, ref = "Teff")
write.csv(counts(dds, normalized = TRUE),
          file = here(analysis_out, "normalized_counts.csv"),
          row.names = FALSE)
```

```{r}
res <- results(dds, name ="Type_Trm_like_vs_Teff")
```

#### Log fold change shrinkage for visualization
```{r}
resultsNames(dds)
```

```{r}
resLFC <- lfcShrink(dds, coef="Type_Trm_like_vs_Teff", type="apeglm")
```
## Annotate and process the results 
```{r}
#### Reformat the gene names ####
# Preserve only the first part of the gene IDs
genes <- rownames(dds)
# Create a table that converts gene names from Ensembl ID
ensembl <- useEnsembl('genes', dataset = 'mmusculus_gene_ensembl',version = 111) # We used release 111 here
annot_raw <- getBM(
  attributes = c(
    'external_gene_name',
    'ensembl_gene_id'),
  filters = 'ensembl_gene_id',
  values = genes,
  mart = ensembl)
annot <- merge(
  x = as.data.frame(genes),
  y =  annot_raw,
  by.y = 'ensembl_gene_id',
  all.x = T,
  by.x = 'genes')
annot <- tibble::column_to_rownames(annot, var = "genes")

saveRDS(annot, file = here(analysis_out, "annot.rds"))
```

```{r}
# Read the annotation file
#annot <- readRDS(here(analysis_out, "annot.rds"))
# Annotate the res obj
res$gene_name <- annot[rownames(res), "external_gene_name"]
res$gene_name_original <- rownames(res)
rownames(res) <- res$gen_name

# Do the same for res_LFC
resLFC$gene_name <- annot[rownames(resLFC), "external_gene_name"]
resLFC$gene_name_original <- rownames(resLFC)
rownames(resLFC) <- resLFC$gen_name
```

```{r}
# Convert the result into a data frame
res_df <- as.data.frame(res)
res_df <- res_df %>% arrange(pvalue)
write.csv(res_df, file = here(analysis_out, "DEG_results.csv"))
```

## Vissualize the results
### Set colors
```{r}
col1 <- "#ae0000"
col2 <- "#095786"
```
### MA Plots
This plot highlights genes with adjusted p-value < 0.1 (blue)
```{r}
DESeq2::plotMA(resLFC, ylim = c(-5, 5), main = "apeglm")
DESeq2::plotMA(res, ylim=c(-5,5), main = "unshrunken")
```
### Plot the normalized counts of genes
First we need to convert the Ensembl IDs into gene names(symbols)
We begin by plotting the CD103 gene (Itgae).
```{r}
plotCounts(dds, gene = "ENSMUSG00000005947", intgroup = "Type", main = res[which(res$gene_name_original=="ENSMUSG00000005947"),"gene_name"])
```
### PCA plot
First we remove the batch effect.
```{r}
# Follow this code for removing batch effect https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#theory 
vsd <- DESeq2::vst(dds, blind = FALSE)
mat <- assay(vsd)
mm <- model.matrix(~Type, colData(vsd))
mat <- limma::removeBatchEffect(mat, batch=vsd$Batch, design=mm)
assay(vsd) <- mat
```
One sample of Trm is an outlier, but we will keep it for now.
```{r}
pca_data <- DESeq2::plotPCA(vsd,intgroup = "Type",returnData = TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"))
pca_plot <- ggplot(pca_data, aes(x = PC1, y = PC2, color = Type)) +
    scale_color_manual(values = c(col2, col1),
                     labels = c(expression(T[Eff]),expression(T[rm]-like)))+
  geom_point(size = 3) +
  theme_minimal() +
  labs(title = "Principal Component Analysis") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  guides(color=guide_legend(""))

pca_plot
```
Or we can try another way. Note here we use the already sorted res_df to get the top genes (ordered by pvalue), and plot the samples. 
```{r}
top_n <- 1000
res.pca <- prcomp(t(assay(vsd)[res_df$gene_name_original[1:top_n],]+1),scale = TRUE)
pca_samples <- fviz_pca_ind(res.pca,
                  #element = "ind",
                  geom = "point",
                  habillage = meta_data$Type,
                  repel = TRUE,
                  palette = c(col2, col1),
                  invisible = "quali",
                  col.ind = meta_data$Type,
                  addEllipses = FALSE,
                  title = paste0('PCA Using Top ',top_n, ' Genes')) + 
  guides(color = guide_legend(override.aes = aes(label = "")))

pca_sample2 <- pca_samples+xlab(str_replace(pca_samples$labels$x, "Dim1", "PC1"))+
  ylab(str_replace(pca_samples$labels$y, "Dim2", "PC2"))
#      theme(legend.position="none")
pca_sample2$layers[[1]]$aes_params$size <- 4 # Change size 
pca_sample2$layers[[1]]$aes_params$alpha <- 0.7 # Change transparency
ggsave(pca_sample2, file = here(analysis_out, "plots", "pca_sample.png"), width = 4,height = 2.6, bg ="white")

pca_sample2
```
### Heatmaps
Delete row and col clusters.
```{r}
# First generate a sample distance matrix
sampleDists <- dist(t(assay(vsd)))
sampleSimMatrix <- max(sampleDists)-as.matrix(sampleDists)

# Set color panel

# Plot heat map
heatmap_samples1 <- pheatmap(sampleSimMatrix,
                             cluster_rows = FALSE,cluster_cols = FALSE,
#                             clustering_distance_rows = sampleDists,
#                             clustering_distance_cols = sampleDists, 
                             col = colorRampPalette(c("white",col1))(75),
                             na_col = "white",
                             border_color = "white",
#                            annotation_col = dplyr::select(meta_data,Type),
#                          annotation_colors = list(Type = c(Trm_like = col2, Teff=col1)),
                          cex = 0.9)

heatmap_samples2 <- pheatmap(sampleSimMatrix,
                             cluster_rows = FALSE,cluster_cols = FALSE,
#                             clustering_distance_rows = sampleDists,
#                             clustering_distance_cols = sampleDists, 
                             col = colorRampPalette(c("blue",col1))(75),
                             na_col = "white",
                             border_color = "white",
#                            annotation_col = dplyr::select(meta_data,Type),
#                          annotation_colors = list(Type = c(Trm_like = col2, Teff=col1)),
                          cex = 0.9)

ggsave(heatmap_samples1, file = here(analysis_out, "plots", "heatmap_samples1.png"), width = 6,height = 6, bg ="white")
ggsave(heatmap_samples2, file = here(analysis_out, "plots", "heatmap_samples2.png"), width = 6,height = 6, bg ="white")
heatmap_samples1
heatmap_samples2
```
### Hierarchical clustering
```{r}
hc <- hclust(sampleDists)
dend <- as.dendrogram(hc)
par(mar=c(1,1,1,5))
dend %>%
  set("labels_col", value = c(col1, col2), k = 2)%>%
  set("branches_k_color", value = c(col1,col2),k = 2) %>%
  plot(horiz = TRUE, axes=FALSE)
```
### Heatmap of top genes
We plot with log transformed normalized counts
```{r}
rld <- rlog(dds, blind = FALSE)
rownames(rld) <- annot[rownames(rld), "external_gene_name"]
```
We can use vsd to plot the gene expression, as we have removed batch effect using limma on assay(vsd).
```{r}
vsd2 <- vsd
rownames(vsd2) <- annot[rownames(vsd2), "external_gene_name"]
top_genes <- res_df[1:50,]$gene_name
vsd_top <- assay(vsd2)[top_genes,]
# Note we have to italicize the genes using bquote()
ita_names <- lapply(rownames(vsd_top), function(x) bquote(italic(.(x))))

par(mar = c(10,1,1,10))
heatmap_genes_top_1 <- pheatmap(vsd_top, cluster_rows = TRUE, cluster_cols = FALSE, 
                              show_rownames = TRUE, show_colnames = TRUE, 
                              scale = "row", main = "Heatmap of top genes",
                          color = colorRampPalette(c("white","#bb180c"))(225),
                          legend_breaks = c(-0.8,-0.4,0,0.4,0.8,max(rld_top)),
                          labels_row = as.expression(ita_names),
#                          annotation_col = dplyr::select(meta_data,Type),
#                          annotation_colors = list(Type = c(Trm_like = col2, Teff=col1)),
                          cex = 0.8)
heatmap_genes_top_2 <- pheatmap(vsd_top, cluster_rows = TRUE, cluster_cols = FALSE,
                              show_rownames = TRUE, show_colnames = TRUE, 
                              scale = "row", main = "Heatmap of top genes",
                          color = colorRampPalette(c(col2, "#f7f7f7",col1 ))(225),
                          legend_breaks = c(-0.8,-0.4,0,0.4,0.8,max(rld_top)),
                          labels_row = as.expression(ita_names),
#                          annotation_col = dplyr::select(meta_data,Type),
                          #annotation_colors = list(Type = c(Trm_like = col2, Teff=col1)),
                          cex = 0.8)
ggsave(heatmap_genes_top_1, file = here(analysis_out, "plots", "heatmap_genes_top_1.png"), width = 5,height = 6)
ggsave(heatmap_genes_top_2, file = here(analysis_out, "plots", "heatmap_genes_top_2.png"), width = 5,height = 6)

heatmap_genes_top_1
heatmap_genes_top_2
```
### Heatmap of specific genes
Genes are defined in this paper [here](https://www.nature.com/articles/ni.2744). We plot with log transformed normalized counts.
```{r}
rld <- rlog(dds, blind = FALSE)
rownames(rld) <- annot[rownames(rld), "external_gene_name"]
```
```{r}
# Define the genes that we are interested in
gene_int <- unique(c("Zfp683","Nur77","Cdh1","Cd244","LOC100503878","Nr4a1","Itga1","LOC641050_1","Skil","Sik1","Litaf","Chn2","Ctla4",
              "LOC641050_2","Icos","Inpp4b","Tmem124","Vps37b","Nr4a2","Nr4a1","Itgae","Rgs1",
              "Xcl1","Rgs2","Hspa1a_1","Hspa1a_2","Hspa1a_3","S1pr1","Icos","Hpgds",
              "S1pr5","Klre1","Elovl7","Eomes","Fgf13", "Prdm1","Usp33","Slamf6","Tlr1",
              "Qpct","Hpgds","Sidt1","Rasgrp2","S1pr1","Cmah","Fam65b","Ly6c2","Hspa1a","Tmem123","Ripor2","S1pr5","Cd224a"))

table(gene_int %in% res_df $gene_name)
```
```{r}
# Plot the heatmap
res_sub <- subset(res_df, gene_name %in% gene_int)
vsd_int <- assay(vsd2)[res_sub$gene_name,]
# Note we have to italicize the genes using bquote()
ita_names <- lapply(rownames(vsd_int), function(x) bquote(italic(.(x))))

heatmap_genes_int1 <- pheatmap(vsd_int, cluster_rows = TRUE, cluster_cols = FALSE, 
                               show_rownames = TRUE, show_colnames = TRUE, 
                               scale = "row", main = "Heatmap of signature genes",
                          color = colorRampPalette(c("white","#bb180c" ))(225),
                          legend_breaks = c(-1.5,0,1.5),
                          labels_row = as.expression(ita_names),
#                          annotation_col = dplyr::select(meta_data,Type),
#                          annotation_colors = list(Type = c(Trm_like = col2, Teff=col1)),
                          cex = 0.8)
heatmap_genes_int2 <- pheatmap(vsd_int, cluster_rows = TRUE, cluster_cols = FALSE, 
                               show_rownames = TRUE, show_colnames = TRUE, 
                               scale = "row", main = "Heatmap of signature genes",
                          # color = colorRampPalette(c("#3b5998", "#f7f7f7","#bb180c" ))(225),
                          color = colorRampPalette(c("#1a1aff", "#EEEEEE","#DA1212" ))(225),
                          legend_breaks = c(-1.5,0,1.5),
                          labels_row = as.expression(ita_names),
#                          annotation_col = dplyr::select(meta_data,Type),
#                          annotation_colors = list(Type = c(Trm_like = col2, Teff=col1)),
                          cex = 0.8)

ggsave(heatmap_genes_int1, file = here(analysis_out,"plots", "heatmap_genes_list1.png"), width = 5,height = 6)
ggsave(heatmap_genes_int2, file = here(analysis_out,"plots", "heatmap_genes_list2.png"), width = 5,height = 6)

heatmap_genes_int1
heatmap_genes_int2
```
### PCA plots of the list of genes
```{r}
norm_counts <- DESeq2::counts(dds, normalized=TRUE)
rownames(norm_counts) <- annot[rownames(norm_counts), "external_gene_name"]
gene_int_2 <- gene_int[gene_int %in% rownames(norm_counts)]
res.pca.genes <- prcomp(t(assay(vsd2)[gene_int_2,]+1),scale = TRUE)
pca_genes <- fviz_pca_biplot(res.pca.genes,
                  #element = "ind",
                  habillage = meta_data$Type,
                  col.var = "#B696B6",
                  repel = TRUE,
                  palette = c(col1, col2),
                  invisible = "quali",
                  col.ind = meta_data$Type,
                  addEllipses = FALSE,
                  title = paste0('Principal Component Analysis with Signature Genes')) + 
  guides(color = guide_legend(override.aes = aes(label = "")))

pca_genes
ggsave(pca_genes, file = here(analysis_out, "plots","pca_genes_list.png"), width = 6,height = 6, bg = "white")
```
### Volcano plot
```{r}
# Create a data frame and label the genes
resLFC_df <- as.data.frame(resLFC)
# Label the genes according to padj threshold = 0.05 and log2 fold change of 0.58 (1.5x fold-change)
padj_threshold <- 0.01
resLFC_df$DE <- "Not Sig"
resLFC_df$DE[resLFC_df$padj<padj_threshold & resLFC_df$log2FoldChange>0.58] <- "Up"
resLFC_df$DE[resLFC_df$padj<padj_threshold & resLFC_df$log2FoldChange< -0.58] <- "Down"
resLFC_df$DE <- factor(resLFC_df$DE, levels = c( "Up","Not Sig", "Down"))
```
```{r}
selected_genes <- resLFC_df%>% 
  subset( abs(log2FoldChange)>3.5 & abs(log10(pvalue))>30 )%>%
  mutate(change = ifelse(log2FoldChange > 0 , "Upregulated", "Downregulated"))

volcano_plot <- ggplot() +
  geom_point(data = resLFC_df, aes(x = log2FoldChange, y = -log10(pvalue), color = DE), alpha = 0.6, shape =16) +
  scale_color_manual(values = c("#ae0000","grey","#095786")) +
  theme_minimal() +
  geom_text_repel(data = selected_genes, aes(x = log2FoldChange, y = -log10(pvalue), label = gene_name, color = DE), 
                  point.padding = 0,
                  force = 0.1, show.legend = FALSE)+
  labs(color = "")+
  labs(x = "Log2 Fold Change", y = "-log10(p-value)") +
  geom_hline(yintercept = -log10(padj_threshold), linetype = "dashed", color = "grey") +
  geom_vline(xintercept = c(-0.58, 0.58), linetype = "dashed", color = "grey")+
  theme(text=element_text(size = 15))
ggsave(volcano_plot, file = here(analysis_out, "plots", "volcano_plot.png"), 
       width = 6,height = 6,bg = "white")

volcano_plot
```

## Pathway analysis
### Enrichment analysis of upregulated genes using GO
For a reference, we can follow the doc here [here](https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html#gene-ontology-analysis).
```{r}
# Select top genes according to p-value and base expression
summary(res$baseMean)
# Select the top upregulated genes that are at least 1.5x more expressed
sig <- res %>% na.omit() 
sig_up <- sig[(sig$padj < 0.05 & sig$baseMean > 10 & sig$log2FoldChange > 0.58),]
sig_up_genes <- sig_up$gene_name_original
```
```{r}
GO_up_results <- enrichGO(gene = sig_up_genes,
                       OrgDb = org.Mm.eg.db,
                       keyType = "ENSEMBL",
                       ont = "BP",
                       pvalueCutoff = 0.05,
                       qvalueCutoff = 0.05,
                       universe = res_df$gene_name_original,
                       readable = TRUE)
s_go_up <-clusterProfiler::simplify(GO_up_results) # Remove redundant genes
GO_res_up_df <- as.data.frame(s_go_up)
write.csv(GO_res_up_df, file = here(analysis_out, "GO_results_up.csv"))
```

```{r}
# Plot GO results with dotplot
plot_GO_up <- dotplot(s_go_up, x = "Count",
                      showCategory = 30, font.size = 8)+
  labs(title = "GO Enrichment Top Up-Regulated Pathways")+theme_minimal()
ggsave(plot_GO_up, file = here(analysis_out, "plots", "GO_up_plot.png"), width = 6,height = 10, bg = "white", dpi = 600)
plot_GO_up
```
### Enrichment analysis of downregulated genes using GO
```{r}
# Select top downregulated genes according to p-value and base expression
summary(res$baseMean)
# Select the top downregulated genes that are at least 2x less expressed
sig <- res %>% na.omit() 
sig_down <- sig[(sig$padj < 0.05 & sig$baseMean > 10 & sig$log2FoldChange < -0.58),]
sig_down_genes <- sig_down$gene_name_original
```
```{r}
GO_down_results <- enrichGO(gene = sig_down_genes,
                       OrgDb = org.Mm.eg.db,
                       keyType = "ENSEMBL",
                       ont = "BP",
                       pvalueCutoff = 0.05,
                       qvalueCutoff = 0.05,
                       universe = res_df$gene_name_original,
                       readable = TRUE)
s_go_down <-clusterProfiler::simplify(GO_down_results) # Remove redundant genes
GO_res_down_df <- as.data.frame(s_go_down)
write.csv(GO_res_down_df, file = here(analysis_out, "GO_results_down.csv"))
```

```{r}
# Plot GO results with dotplot
plot_GO_down <- dotplot(s_go_down, x = "Count",
                        showCategory = 30, font.size = 8)+
  labs(title = "GO Enrichment Top Down-Regulated Pathways")+theme_minimal()

ggsave(plot_GO_down, file = here(analysis_out, "plots", "GO_down_plot.png"), width = 6,height = 10, bg = "white", dpi = 600)
plot_GO_down
```
### Cnetplot
```{r}
FC_up <- sig_up$log2FoldChange
FC_down <- sig_down$log2FoldChange

names(FC_up) <- sig_up$gene_name_original
names(FC_down) <- sig_down$gene_name_original

# We look at selected pathways and visualize the relationships between the genes 
s_go_up_sel <- s_go_up
s_go_down_sel <- s_go_down

s_go_up_sel@result <- s_go_up[c(1,3,5,6),]
s_go_down_sel@result <- s_go_down[c(1,3,5),]

cnet_up_1 <-cnetplot(s_go_up_sel,
                     categorySize = "pvalue",
         shadowtext='gene',  
         cex.params = list(gene_label = 0.4, category_label = 0.8),
         color.params = list(category = "red", 
                             foldChange =  FC_up))+
  scale_color_continuous(name = "Fold Change", low = "blue", high = "red")
cnet_down_1 <-cnetplot(s_go_down_sel,
                     categorySize = "pvalue",
         shadowtext='gene',  
         cex.params = list(gene_label = 0.4, category_label = 0.8),
         color.params = list(category = "blue", 
                             foldChange =  FC_down))+
  scale_color_continuous(name = "Fold Change", low = "blue", high = "lightblue")

ggsave(cnet_up_1, file = here(analysis_out, "plots", "cnet_up_1.png"), width = 10,height = 6, bg = "white")
ggsave(cnet_down_1, file = here(analysis_out, "plots", "cnet_down_1.png"), width = 10,height = 6, bg = "white")
cnet_up_1
cnet_down_1
```
### Tree Plot
```{r}
#semdata <- GOSemSim::godata('org.Mm.eg.db', ont = "BP")
#s_go_up <- pairwise_termsim(s_go_up, semData = semdata, method = "JC")
s_go_up <- pairwise_termsim(s_go_up)
s_go_down <- pairwise_termsim(s_go_down)

tree_up <- suppressMessages(treeplot(s_go_up,fontsize = 3,
         cluster.params = list(method = "ward.D",
                               n = 5,
                               label_format = 30,
                               color = rep("black",5)),
         hilight.params = list(hilight = FALSE)))+
  ggtitle("Up-Regulated Pathways")

tree_down <- suppressMessages(treeplot(s_go_down,fontsize = 3,
         cluster.params = list(method = "ward.D",
                               n = 5,
                               label_format = 30,
                               color = rep("black",5)),
         hilight.params = list(hilight = FALSE)))+
  ggtitle("Down-Regulated Pathways")

ggsave(tree_up, file = here(analysis_out, "plots", "tree_up.png"), width = 12,height = 6, bg = "white")
ggsave(tree_down, file = here(analysis_out, "plots", "tree_down.png"), width = 12,height = 6, bg = "white")
tree_up
tree_down
```
### Condition comparison
```{r}
comparelist <- list(sig_up$gene_name_original, sig_down$gene_name_original)
names(comparelist) <- c("Upregulated", "Downregulated")
cclust<-compareCluster(geneCluster = comparelist, 
               fun = enrichGO,
               OrgDb= org.Mm.eg.db,
               keyType = "ENSEMBL",
               ont= "BP",
               universe=res_df$gene_name_original)
```
```{r}
compareplot <- dotplot(cclust, showCategory = 10, font.size = 10, by = "Count")+
  xlab("")
ggsave(compareplot, file = here(analysis_out, "plots", "compare_plot.png"), width = 6,height = 7, bg = "white")
compareplot
```

### GSEA analysis
Curate a gene list
```{r}
res_df_filt <- res_df %>% filter(!is.na(padj)&baseMean>10)%>%
  arrange(desc(log2FoldChange))
# Create an ordered list of gene foldchanges
gene_list <- res_df_filt$log2FoldChange
names(gene_list) <- res_df_filt$gene_name_original
```
```{r}
gse <- gseGO(geneList = gene_list,
             OrgDb = org.Mm.eg.db,
             ont = "BP",
             keyType = "ENSEMBL",
             minGSSize = 10,
             pvalueCutoff = 0.05,
             verbose = TRUE)
s_gse <- clusterProfiler::simplify(gse) # Simplify and remove redundant terms
```
```{r}
gse_res <- s_gse@result
gse_res <- gse_res[order(gse_res$NES),]

write.csv(gse_res, file = here(analysis_out, "GSEA_results.csv"))
```

Plot pathways
```{r}
geneset_up_selected <- c(73,75,76,83,84,102)
geneset_down_selected <- c(61,64,65,66,67)

gsea_down_1 <- gseaplot2(s_gse, geneSetID = gse_res$ID[geneset_down_selected], pvalue_table = TRUE,
                         title = "Down-Regulated Pathways GSEA",
                        base_size = 11,
                        color = c("#264653","#2a9d8f", "#e9c46a", "#f4a261", "#e76f51"))


gsea_up_1 <- gseaplot2(s_gse, geneSetID = gse_res$ID[geneset_up_selected], 
                       pvalue_table = TRUE,
                       title = "Up-Regulated Pathways GSEA",
                       base_size = 11,
                       color = c("#264653","#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","orange"))


ggsave(gsea_down_1, file = here(analysis_out, "plots", "GSEA_down_1.png"), width = 13,height = 8, bg = "white")
ggsave(gsea_up_1, file = here(analysis_out, "plots", "GSEA_up_1.png"), width = 13,height = 8, bg = "white")

gsea_down_1
gsea_up_1

```

